### PR TITLE
Implement navigation events

### LINKED ISSUE #11005: Implement navigation events
I am working on this

Part of #10992 

Events:
 * `PopStateEvent`
 * `HashChangeEvent`
 * `PageTransitionEvent`

Specs:
https://html.spec.whatwg.org/multipage/browsers.html#the-popstateevent-interface 
https://html.spec.whatwg.org/multipage/browsers.html#hashchangeevent 
https://html.spec.whatwg.org/multipage/browsers.html#pagetransitionevent

### PR DESCRIPTION
**Pull request by cbrewster**
fixes #11005

Events:
- `PopStateEvent`
- `HashChangeEvent`
- `PageTransitionEvent`

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11006)

### CODE REVIEW

**Comment by ConnorGBrewster** [comment]
This is ready for review!

File: components/script/dom/hashchangeevent.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+               -> Root<HashChangeEvent> {
+        let ev = HashChangeEvent::new_uninitialized(global, old_url, new_url);
+        {
+            let event = ev.upcast::<Event>();
+            event.init_event(type_, bubbles, cancelable);
+        }
+        ev
+    }
+
+    #[allow(unsafe_code)]
```
Comment (jdm) [inline components/script/dom/hashchangeevent.rs]: Eh?

Comment (ConnorGBrewster) [inline components/script/dom/hashchangeevent.rs]: whoops! there should only be on the PopStateEvent due to the webidl `any`. But I will remove it on the other 2 events.

**Comment by asajeffrey** [comment]
From the test results:
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface PageTransitionEvent to all webpages as a property on the window? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:251:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:245:1
```
@jdm or @Ms2ger you have been named!

**Comment by asajeffrey** [comment]
Looks good. You can r=me once @jdm or @Ms2ger have blessed it.

<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> From the test results:
> 
> ```
> 
>   ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
> 
>   │ FAIL [expected PASS] Interfaces exposed on the window
> 
>   │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface PageTransitionEvent to all webpages as a property on the window? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
> 
>   │ 
> 
>   │ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:251:5
> 
>   │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
> 
>   │ test@http://web-platform.test:8000/resources/testharness.js:495:9
> 
>   └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:245:1
> 
> ```
> 
> @jdm or @Ms2ger you have been named!</details>


---

Reviewed 6 of 7 files at r1, 1 of 1 files at r2, 2 of 4 files at r3, 1 of 1 files at r4, 1 of 2 files at r9, 1 of 1 files at r11.
Review status: 7 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/popstateevent.rs, line 24 \[r11\]](https://reviewable.io:443/reviews/servo/servo/11006#-KHVi_0j0-rhdG_D39FT:-KHVi_0j0-rhdG_D39FU:1171629304) ([raw file](https://github.com/servo/servo/blob/aef177fed2069db1c6a73583e3778ead793a4257/components/script/dom/popstateevent.rs#L24)):*
> ```Rust
>     event: Event,
>     #[ignore_heap_size_of = "Defined in rust-mozjs"]
>     state: MutHeapJSVal,
> ```

Pity we can't use a more precise type here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11006#-:-KHViktD-d5nkfMML_Ls:891442272)*

**Comment by ConnorGBrewster** [comment]
Just squashed, that test is passing, I added `PageTransitionEvent` to the exposed properties list after that test failure.

**Comment by jdm** [comment]
That is the correct thing to do.

**Comment by ConnorGBrewster** [comment]
I am unsure why that test is failing. It is not marked as expected to fail.

**Comment by jdm** [comment]
http://mxr.mozilla.org/servo/source/tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/PopStateEvent.html.ini

**Comment by ConnorGBrewster** [comment]
@jdm ahh need to rebase my local version

**Pull request merged**